### PR TITLE
Add fn_handle_safe_ts_update ffi

### PR DESCRIPTION
**Pull request by hehechen**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflash/issues/4902  https://github.com/tikv/tikv/issues/13435

Problem Summary:

### What is changed and how it works?
Add fn_handle_safe_ts_update ffi
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
Add log in TiFlash to pring safe ts.
- [ ] No code

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/proxy_server/src/run.rs [hunk 1]
```diff
@@ -1129,6 +1124,44 @@ impl<ER: RaftEngine> TiKvServer<ER> {
...
+
+        let tiflash_ob = engine_store_ffi::observer::TiFlashObserver::new(
+            node.id(),
+            self.engines.as_ref().unwrap().engines.kv.clone(),
+            importer.clone(),
+            self.proxy_config.raft_store.snap_handle_pool_size,
+        );
+        tiflash_ob.register_to(self.coprocessor_host.as_mut().unwrap());
+
+        let check_leader_runner = CheckLeaderRunner::new(
```
Comment (CalvinNeo) [inline components/proxy_server/src/run.rs:1157]: Why change order?

File: new-mock-engine-store/src/lib.rs [hunk 2]
```diff
@@ -659,6 +659,7 @@ pub fn gen_engine_store_server_helper(
         fn_get_config: None,
         fn_set_store: None,
         fn_set_pb_msg_by_bytes: Some(ffi_set_pb_msg_by_bytes),
+        fn_handle_safe_ts_update: None,
```
Comment (CalvinNeo) [inline new-mock-engine-store/src/lib.rs:662]: If you changed definitions of FFI, we need to `make gen_proxy_ffi` first

File: components/proxy_server/src/run.rs (hunk 1 above)
Comment (hehechen) [reply components/proxy_server/src/run.rs:1157]: Because the observer need to be registered before check_leader runner be created.

**Review by CalvinNeo — APPROVED** [review]
LGTM. Tests will be added in later PR.

**Pull request merged**